s3pypi/core.py

```diff
@@ -58,7 +68,8 @@ def upload_packages(
             index = storage.get_index(directory)
 
             for distr in group:
-                filename = distr.local_path.name
+                filename = distr.filename
+                index.replace_filename_without_hash(filename)
```

Comment:
What happens when a file is replaced using `--force`, but it had a hash previously? It looks like the filename will be duplicated in the index, one for each hash.

Reply:
I believe the `replace_filename_without_hash` method should be called `put` instead, and it should overwrite any existing filename whether it had a hash or not. The `if not force` below will protect from accidental overwrites.

Reply:
Given this new behaviour, it might make sense to turn `Index.filenames` into a dict. Keeping it a set (and overriding `__hash__`) is confusing.

Reply:
Yea, using `--force` did duplicate a package in the index. 

I've made changes following your advice. The behavior looks clearer and predictable now